File: parabolic/example/prep_grid/model2.m
% creat boundary data file bz1 bz2

clc;
clear all;
close all;

flag_printf = 1;

nx = 401;
ny = 301;
dh = 10;
origin_x = 0;
origin_y = 0;
origin_z = 0;
bz1 = zeros(nx,ny,3);
bz2 = zeros(nx,ny,3);

p1 = 1;
p2 = 101;
p3 = 151;
p4 = 251;
p5 = 301;
p6 = 401;

for j=1:ny
  for i=p1:p2
    bz2(i,j,1) = origin_x + (i-1)*dh;
    bz2(i,j,2) = origin_y + (j-1)*dh;
    bz2(i,j,3) = origin_z;
  end
  for i=p2+1:p3
    bz2(i,j,1) = origin_x + bz2(p2,j,1) + cos(0.5*pi*70/90)*(i-p2)*dh;
    bz2(i,j,2) = origin_y + (j-1)*dh;
    bz2(i,j,3) = origin_z + bz2(p2,j,3) - sin(0.5*pi*70/90)*(i-p2)*dh;
  end
  for i=p3+1:p4
    bz2(i,j,1) = origin_x + bz2(p3,j,1) + (i-p3)*dh;
    bz2(i,j,2) = origin_y + (j-1)*dh;
    bz2(i,j,3) = origin_z + bz2(p3,j,3);
  end
  for i=p4+1:p5
    bz2(i,j,1) = origin_x + bz2(p4,j,1) + cos(0.5*pi*70/90)*(i-p4)*dh;
    bz2(i,j,2) = origin_y + (j-1)*dh;
    bz2(i,j,3) = origin_z + bz2(p4,j,3) + sin(0.5*pi*70/90)*(i-p4)*dh;
  end
  for i=p5+1:p6
    bz2(i,j,1) = origin_x + bz2(p5,j,1) + (i-p5)*dh;
    bz2(i,j,2) = origin_y + (j-1)*dh;
    bz2(i,j,3) = origin_z + bz2(p5,j,3);
  end
end

% smooth alone two direction
for i=1:nx
  bz2(i,:,1) = smooth(bz2(i,:,1),40);
  bz2(i,:,2) = smooth(bz2(i,:,2),40);
  bz2(i,:,3) = smooth(bz2(i,:,3),40);
end

for j=1:ny
 bz2(:,j,1) = smooth(bz2(:,j,1),40);
 bz2(:,j,2) = smooth(bz2(:,j,2),40);
 bz2(:,j,3) = smooth(bz2(:,j,3),40);
end

min(bz2(:,1,1))
max(bz2(:,1,1))
% min(bz2(:,1,2))
% max(bz2(:,1,2))
x=linspace(-500,3800,401);
y=linspace(0,3000,301);
for i=1:nx
  for j=1:ny
    bz1(i,j,1) = x(i);
    bz1(i,j,2) = y(j);
    bz1(i,j,3) = -2000;
  end
end


% A = 0.000001;
% [bz2] = arc_stretch_x(A,bz2);


figure(1)
surf(bz2(:,:,1)/1e3,bz2(:,:,2)/1e3,bz2(:,:,3)/1e3);
shading interp;
hold on;
plot3(bz2(:,151,1)/1e3,bz2(:,151,2)/1e3,bz2(:,151,3)/1e3,'r',LineWidth=2);

plot3(1.0, 0.5, +0.0, 'rv',MarkerSize=20,LineWidth=2);
plot3(1.0, 1.5, +0.0, 'rv',MarkerSize=20,LineWidth=2);
plot3(1.0, 2.5, +0.0, 'rv',MarkerSize=20,LineWidth=2);
plot3(1.5, 1.5, -0.4, 'rv',MarkerSize=20,LineWidth=2);
plot3(1.5, 0.5, -0.4, 'rv',MarkerSize=20,LineWidth=2);
text(1.0, 0.5-0.1, +0.0+0.2,'R1',Color='r',FontSize=20,FontWeight='bold');
text(1.0, 1.5-0.1, +0.0+0.2,'R2',Color='r',FontSize=20,FontWeight='bold');
text(1.0, 2.5-0.1, +0.0+0.2,'R3',Color='r',FontSize=20,FontWeight='bold');
text(1.5, 1.5-0.1, -0.4+0.2,'R4',Color='r',FontSize=20,FontWeight='bold');
text(1.5, 0.5-0.1, -0.4+0.2,'R5',Color='r',FontSize=20,FontWeight='bold');
plot3(1,1,0.0,'rp',MarkerSize=20,LineWidth=2);

grid off;
axis equal;
% caxis([0,0.6]);
cid = colorbar;
cid.Label.String='(km)';
%
set(gcf,'color','white');
set(gcf,'Position',[0,0,2200,1000]);
xlabel('X axis (km)',FontWeight='bold',rotation=-25,position=[2.5 -1.3]);
ylabel('Y axis (km)',FontWeight='bold',rotation=40, position=[4.5 0.5]);
zlabel('Z axis (km)',FontWeight='bold',rotation=0,position=[-0.3 0.3]);
set(gca,'layer','top');
set(gca, FontWeight='bold', FontSize=15);
view(40,30);

% creat data file
file_name = '../data_file_3d.txt';
fid=fopen(file_name,'w'); % Output file name 
fprintf(fid,'# nx ny number\n'); 
fprintf(fid,'%d %d\n',nx,ny);
fprintf(fid,'# bz1 coords\n'); 
for j=1:ny
  for i=1:nx
    fprintf(fid,'%.9e %.9e %.9e\n',bz1(i,j,1),bz1(i,j,2),bz1(i,j,3));
  end
end
fprintf(fid,'# bz2 coords\n'); 
for j=1:ny
  for i=1:nx
    fprintf(fid,'%.9e %.9e %.9e\n',bz2(i,j,1),bz2(i,j,2),bz2(i,j,3));
  end
end

if flag_printf
   print(gcf,'model2.png','-r400','-dpng');
end

File: parabolic/example/prep_grid/model2.py
import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', '..', 'plotting'))
from export_c import export_bdry_2

import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt


def smooth_1d(arr, span):
    """Reimplementation of MATLAB smooth(y, span) with 'moving' method.

    Even span is reduced by 1 (MATLAB convention). At boundaries, uses
    anchored windows (left-anchored at start, right-anchored at end)
    rather than centered truncated windows.
    """
    arr = np.asarray(arr, dtype=float)
    n = len(arr)
    if span % 2 == 0:
        span = span - 1
    half = span // 2
    result = np.zeros(n)
    for i in range(n):
        if i < half:
            lo = 0
            hi = 2 * i
        elif i >= n - half:
            lo = 2 * i - n + 1
            hi = n - 1
        else:
            lo = i - half
            hi = i + half
        result[i] = np.mean(arr[lo:hi+1])
    return result


def main():
    flag_printf = 1

    nx = 401
    ny = 301
    dh = 10
    origin_x = 0
    origin_y = 0
    origin_z = 0

    bz1 = np.zeros((nx, ny, 3))
    bz2 = np.zeros((nx, ny, 3))

    # Convert to 0-based indices
    p1 = 0
    p2 = 100
    p3 = 150
    p4 = 250
    p5 = 300
    p6 = 400

    for j in range(ny):
        # i from p1 to p2 (inclusive)
        for i in range(p1, p2 + 1):
            bz2[i, j, 0] = origin_x + i * dh
            bz2[i, j, 1] = origin_y + j * dh
            bz2[i, j, 2] = origin_z
        # i from p2+1 to p3
        for i in range(p2 + 1, p3 + 1):
            bz2[i, j, 0] = origin_x + bz2[p2, j, 0] + np.cos(0.5 * np.pi * 70 / 90) * (i - p2) * dh
            bz2[i, j, 1] = origin_y + j * dh
            bz2[i, j, 2] = origin_z + bz2[p2, j, 2] - np.sin(0.5 * np.pi * 70 / 90) * (i - p2) * dh
        # i from p3+1 to p4
        for i in range(p3 + 1, p4 + 1):
            bz2[i, j, 0] = origin_x + bz2[p3, j, 0] + (i - p3) * dh
            bz2[i, j, 1] = origin_y + j * dh
            bz2[i, j, 2] = origin_z + bz2[p3, j, 2]
        # i from p4+1 to p5
        for i in range(p4 + 1, p5 + 1):
            bz2[i, j, 0] = origin_x + bz2[p4, j, 0] + np.cos(0.5 * np.pi * 70 / 90) * (i - p4) * dh
            bz2[i, j, 1] = origin_y + j * dh
            bz2[i, j, 2] = origin_z + bz2[p4, j, 2] + np.sin(0.5 * np.pi * 70 / 90) * (i - p4) * dh
        # i from p5+1 to p6
        for i in range(p5 + 1, p6 + 1):
            bz2[i, j, 0] = origin_x + bz2[p5, j, 0] + (i - p5) * dh
            bz2[i, j, 1] = origin_y + j * dh
            bz2[i, j, 2] = origin_z + bz2[p5, j, 2]

    # Smooth along two directions
    window_len = 40
    for i in range(nx):
        bz2[i, :, 0] = smooth_1d(bz2[i, :, 0], window_len)
        bz2[i, :, 1] = smooth_1d(bz2[i, :, 1], window_len)
        bz2[i, :, 2] = smooth_1d(bz2[i, :, 2], window_len)
    for j in range(ny):
        bz2[:, j, 0] = smooth_1d(bz2[:, j, 0], window_len)
        bz2[:, j, 1] = smooth_1d(bz2[:, j, 1], window_len)
        bz2[:, j, 2] = smooth_1d(bz2[:, j, 2], window_len)

    # bz1 x: linspace(-500, 3800, 401), y: linspace(0, 3000, 301), z=-2000
    x = np.linspace(-500, 3800, nx)
    y = np.linspace(0, 3000, ny)
    for j in range(ny):
        for i in range(nx):
            bz1[i, j, 0] = x[i]
            bz1[i, j, 1] = y[j]
            bz1[i, j, 2] = -2000

    file_name = '../data_file_3d.txt'
    export_bdry_2(bz1, bz2, nx, ny, file_name)

    if flag_printf:
        os.makedirs('../plot', exist_ok=True)
        stride = max(1, max(nx, ny) // 80)
        fig = plt.figure(figsize=(11, 6))
        ax = fig.add_subplot(111, projection='3d')
        X = bz2[:, :, 0] / 1e3; Y = bz2[:, :, 1] / 1e3; Z = bz2[:, :, 2] / 1e3
        ax.plot_surface(X[::stride, ::stride], Y[::stride, ::stride],
                        Z[::stride, ::stride], cmap='jet', alpha=0.9, edgecolor='none')
        mid_j = ny // 2
        ax.plot(X[:, mid_j], Y[:, mid_j], Z[:, mid_j], 'r', linewidth=2)
        ax.set_xlabel('X axis (km)', fontweight='bold')
        ax.set_ylabel('Y axis (km)', fontweight='bold')
        ax.set_zlabel('Z axis (km)', fontweight='bold')
        ax.view_init(elev=30, azim=40)
        ax.set_box_aspect([1, 1, 0.5])
        ax.grid(False)
        mappable = plt.cm.ScalarMappable(cmap='jet')
        mappable.set_array(Z[::stride, ::stride].ravel())
        mappable.set_clim(Z.min(), Z.max())
        cb = fig.colorbar(mappable, ax=ax, shrink=0.6, pad=0.1)
        cb.set_label('(km)', fontweight='bold')
        fig.patch.set_facecolor('white')
        plt.tight_layout()
        plt.savefig('../plot/para_model2.png', dpi=300, bbox_inches='tight', facecolor='white')
        plt.close(fig)
        print('Saved: ../plot/para_model2.png')


if __name__ == "__main__":
    main()
